opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -619,15 +619,15 @@ namespace Opm {
                 PrimaryVariables& priVars = solution[ cell_idx ];
 
                 const double& dp = dx[cell_idx][Indices::pressureSwitchIdx];
-                double& p = priVars[Indices::pressureSwitchIdx];
+                Scalar& p = priVars[Indices::pressureSwitchIdx];
                 const double& dp_rel_max = dpMaxRel();
                 const int sign_dp = dp > 0 ? 1: -1;
                 p -= sign_dp * std::min(std::abs(dp), std::abs(p)*dp_rel_max);
                 p = std::max(p, 0.0);
 
                 // Saturation updates.
-                const double dsw = active_[Water] ? dx[cell_idx][Indices::waterSaturationIdx] : 0.0;
-                const double dxvar = active_[Gas] ? dx[cell_idx][Indices::compositionSwitchIdx] : 0.0;
+                const double dsw = active_[Water] ? dx[cell_idx][Indices::waterSaturationIdx] : Scalar(0);
```

Comment:
Should not the dsw etc. variables also be Scalar? As written this seems to force Scalar to be implicitly convertible to double, which may not be possible? Or is that an assumption that is made also elsewhere?